### ISSUE TITLE
Bound generic constructor parameter doesn't type check if constructor called as an argument to `Any` parameter

### ISSUE THREAD

**Issue by bryanforbes** [microsoft/pyright#3979]
**Describe the bug**
Given the following code:

```python
from __future__ import annotations

from typing import Any, Generic, TypeVar

T = TypeVar('T', bound=str)


class Thing(Generic[T]):
    item: T

    def __init__(self, item: T) -> None:
        self.item = item


def do_something(thing: Any) -> None:
    ...


Thing('asdf')
Thing(1)  # error: "Literal[1]" cannot be assigned to type "T@Thing"

do_something(Thing('asdf'))
do_something(Thing(1))  # no error
```

No error is generated for line 23

**Expected behavior**
The same error generated at line 20 should be generated for line 23

**VS Code extension or command-line**
Are you running pyright as a VS Code extension or a command-line tool?
Command line

Which version?
1.1.272

**Comment by erictraut** [comment]
Yes, this is expected. It's related to bidirectional type inference. When solving for `T` in the context of an `Any`, the expected result is `Thing[Any]`, so all other checks (e.g. for bound compatibility) are subverted.

Mypy has the same behavior in this case.

You should avoid using `Any` if you want to avoid false negatives like this. For example, change the annotation for the `thing` parameter from `Any` to `object`.
